File: languages/go/projects/03-advanced/main.go
package main

import "fmt"

type Course struct {
	title    string
	capacity int
	enrolled int
}

func NewCourse(title string, capacity int) Course {
	if capacity < 0 {
		capacity = 0
	}
	return Course{title: title, capacity: capacity, enrolled: 0}
}

func (c *Course) Enroll() bool {
	if c.enrolled >= c.capacity {
		return false
	}
	c.enrolled++
	return true
}

func (c *Course) Drop() bool {
	if c.enrolled <= 0 {
		return false
	}
	c.enrolled--
	return true
}

func (c Course) PrintStatus() {
	fmt.Printf("Course: %s | %d/%d enrolled\n", c.title, c.enrolled, c.capacity)
}

func main() {
	goBasics := NewCourse("GoBasics", 2)
	algorithms := NewCourse("Algorithms", 3)

	goBasics.Enroll()
	goBasics.Enroll()
	goBasics.Enroll()

	algorithms.Enroll()

	goBasics.PrintStatus()
	algorithms.PrintStatus()
}

File: languages/python/projects/03-advanced/main.py
class Course:
    def __init__(self, title: str, capacity: int) -> None:
        self._title = title
        self._capacity = max(0, capacity)
        self._enrolled = 0

    def enroll(self) -> bool:
        if self._enrolled >= self._capacity:
            return False
        self._enrolled += 1
        return True

    def drop(self) -> bool:
        if self._enrolled <= 0:
            return False
        self._enrolled -= 1
        return True

    def print_status(self) -> None:
        print(f"Course: {self._title} | {self._enrolled}/{self._capacity} enrolled")


def main() -> None:
    python_basics = Course("PythonBasics", 2)
    algorithms = Course("Algorithms", 3)

    python_basics.enroll()
    python_basics.enroll()
    python_basics.enroll()

    algorithms.enroll()

    python_basics.print_status()
    algorithms.print_status()


if __name__ == "__main__":
    main()
